File: week2_algorithmic_warmup/6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.cpp
#include <iostream>
#include<ctime>

int fibonacci_sum_naive(long long n) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;
    long long sum      = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        sum += current;
    }

    return sum % 10;
}

int fibonacci_sum_new(long long n) {
    if (n <= 1)
        return n;

    int previous = 0;
    int current  = 1;
    int sum      = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = (tmp_previous + current)%10;
        sum += current;
    }

    return sum % 10;
}


void test_case(){

    for (int i=0;i<=30;++i){
        // int n = rand()%30 ;
        clock_t start = clock();
        int sol1 = fibonacci_sum_naive(i);
        double elapsed_1 = (double)(clock()-start)*1000000/CLOCKS_PER_SEC;
        int sol2 = fibonacci_sum_new(i);
        double elapsed_2 = (double)(clock()-start-elapsed_1)*1000000/CLOCKS_PER_SEC;
        if (sol1 == sol2) std::cout<<"OK! \n";
        else std::cout<<"WORNG! \n"<<sol1<<"  "<<sol2<<"\n";
        std::cout<<"TIMES: "<< elapsed_1<<" , "<<elapsed_2<<"\n";
    }
}

int main() {
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum_new(n);
    // test_case();
    return 0;
}

File: week2_algorithmic_warmup/6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.py
def fibonacci_sum(n):
    if n <= 1:
        return n

    previous, current, _sum = 0, 1, 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        _sum += current

    return _sum % 10

def fibonacci_sum_new(n):
    if n <= 1:
        return n

    previous, current, _sum = 0, 1, 1

    for _ in range(n - 1):
        previous, current = current, (previous + current)%10 
        _sum += (current)

    return _sum%10 

def list_gen(n):
    f = [0]
    for i in range(1,n):
        f.append(fibonacci_sum_new(i))
    print(f)


def test_cases():
    import time
    for n in range(100000):
        t = time.perf_counter() * 1000
        sol1 = fibonacci_sum(n)
        t1 = time.perf_counter() * 1000- t
        sol2 = fibonacci_sum_new(n)
        t2 = time.perf_counter() * 1000 - t1 - t 
        if sol1 == sol2:
            print("OK!")
        else:
            print("WRONG!")
            break
        
        print("Solution_1: ",sol1," Time: ", t1)
        print("Solution_2: ",sol2," Time: ", t2)





if __name__ == '__main__':
    n = int(input())
    print(fibonacci_sum_new(n))
    # list_gen(n)
    # test_cases()
